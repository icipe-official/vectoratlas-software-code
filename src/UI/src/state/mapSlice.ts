import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { fetchApiJson } from '../api/api';

export interface MapState {
  map_styles: {
    layers: {
      name: string;
      fillColor: number[];
      strokeColor: number[];
      strokeWidth: number;
      zIndex: number;
    }[];
  };

  map_overlays: {
    name: string;
    source: string;
    sourceType: string;
    layers: { name: string }[];
  }[];
}

export const initialState: MapState = {
  map_styles: { layers: [] },
  map_overlays: [],
};

export const getMapStyles = createAsyncThunk('map/getMapStyles', async () => {
  const mapStyles = await fetchApiJson('config/map-styles');
  return mapStyles;
});

export const getTileServerOverlays = createAsyncThunk(
  'map/getTileServerOverlays',
  async () => {
    const tileServerOverlays = await fetchApiJson(
      'config/tile-server-overlays'
    );
    return tileServerOverlays;
  }
);

export const mapSlice = createSlice({
  name: 'map',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getMapStyles.fulfilled, (state, action) => {
        state.map_styles = action.payload;
      })
      .addCase(getTileServerOverlays.fulfilled, (state, action) => {
        state.map_overlays = action.payload;
      });
  },
});

export default mapSlice.reducer;
