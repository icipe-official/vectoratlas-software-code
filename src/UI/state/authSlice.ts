import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import * as njwt from 'njwt';
import { fetchAuth } from '../api/api';

export interface AuthState {
  roles: String[];
  token: String;
}

export const initialState: AuthState = {
  roles: [],
  token: '',
};

export const getUserInfo = createAsyncThunk('auth/getUserInfo', async () => {
  const token = await fetchAuth();
  const verifiedToken: any = njwt.verify(
    token,
    process.env.NEXT_PUBLIC_TOKEN_KEY
  );
  return {
    token: token,
    roles: verifiedToken?.body.scope.split(','),
  };
});

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getUserInfo.fulfilled, (state, action: any) => {
      state.roles = action.payload.roles;
      state.token = action.payload.token;
    });
  },
});

export default authSlice.reducer;
