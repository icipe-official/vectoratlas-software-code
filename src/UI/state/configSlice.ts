import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import {
  fetchLocalVersion,
  fetchApiVersion,
  fetchFeatureFlags,
} from '../api/api';

export interface ConfigState {
  version_ui: string;
  version_api: string;
  feature_flags_status: string;
  feature_flags: {
    flag: string;
    on: boolean;
  }[];
}

export const initialState: ConfigState = {
  version_ui: 'local_ui',
  version_api: 'local_api',
  feature_flags: [],
  feature_flags_status: '',
};

export const getUiVersion = createAsyncThunk(
  'config/getUiVersion',
  async () => {
    const version = await fetchLocalVersion();
    return version;
  }
);

export const getApiVersion = createAsyncThunk(
  'config/getApiVersion',
  async () => {
    const version = await fetchApiVersion();
    return version;
  }
);

export const getFeatureFlags = createAsyncThunk(
  'config/getFeatureFlags',
  async () => {
    const featureFlags = await fetchFeatureFlags();
    return featureFlags;
  }
);

export const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(getUiVersion.pending, (state) => {
        state.version_ui = 'loading';
      })
      .addCase(getUiVersion.rejected, (state) => {
        state.version_ui = 'error';
      })
      .addCase(getUiVersion.fulfilled, (state, action) => {
        state.version_ui = action.payload;
      })
      .addCase(getApiVersion.pending, (state) => {
        state.version_api = 'loading';
      })
      .addCase(getApiVersion.rejected, (state) => {
        state.version_api = 'error';
      })
      .addCase(getApiVersion.fulfilled, (state, action) => {
        state.version_api = action.payload;
      })
      .addCase(getFeatureFlags.pending, (state) => {
        state.feature_flags_status = 'loading';
      })
      .addCase(getFeatureFlags.rejected, (state) => {
        state.feature_flags_status = 'error';
      })
      .addCase(getFeatureFlags.fulfilled, (state, action) => {
        state.feature_flags_status = 'success';
        state.feature_flags = action.payload;
      });
  },
});

export default configSlice.reducer;
