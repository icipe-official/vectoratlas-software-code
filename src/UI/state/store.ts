import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import rootReducer from './rootReducer';

export function makeStore() {
  return configureStore({
    reducer: rootReducer,
  });
}

const store = makeStore();

export type AppState = ReturnType<typeof store.getState>;
export const rootInitialState = store.getState() as AppState;

export type AppDispatch = typeof store.dispatch;

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  AppState,
  unknown,
  Action<string>
>;

export default store;
