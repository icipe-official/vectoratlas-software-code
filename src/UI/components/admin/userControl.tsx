import { IconButton } from '@mui/material';
import Checkbox from '@mui/material/Checkbox';
import Typography from '@mui/material/Typography';
import React, { useState } from 'react';
import {
  saveUserRoles,
} from '../../state/admin/actions/admin.actions';
import { useAppDispatch, useAppSelector } from '../../state/hooks';
import { UsersWithRoles } from '../../state/state.types';
import SaveIcon from '@mui/icons-material/Save';

export const areRolesDifferent = (
  user: UsersWithRoles,
  workingCopy: UsersWithRoles
) => {
  return (
    user.is_admin !== workingCopy.is_admin ||
    user.is_uploader !== workingCopy.is_uploader ||
    user.is_reviewer !== workingCopy.is_reviewer ||
    user.is_editor !== workingCopy.is_editor
  );
};

export const UserControl = ({ user }: { user: UsersWithRoles }) => {
  const dispatch = useAppDispatch();
  const savingRoles = useAppSelector((s) => s.admin.savingUser);

  const [workingCopy, setWorkingCopy] = useState({ ...user });

  const onChange = (propName: string) => (e) => {
    setWorkingCopy({
      ...workingCopy,
      [propName]: e.target.checked,
    });
  };

  const onSave = () => {
    dispatch(saveUserRoles(workingCopy));
  };

  return (
    <div
      style={{
        display: 'flex',
        width: '100%',
        alignItems: 'center',
        marginBottom: '20px',
      }}
    >
      <div style={{ flexGrow: 1, alignItems: 'center' }}>
        <Typography style={{ fontWeight: 'bold', marginRight: '5px' }}>
          {user.email}
        </Typography>
        <Typography variant="caption">({user.auth0_id})</Typography>
      </div>
      <Typography style={{}}>Uploader:</Typography>
      <Checkbox
        checked={workingCopy.is_uploader}
        onChange={onChange('is_uploader')}
      />
      <Typography style={{}}>Reviewer:</Typography>
      <Checkbox
        checked={workingCopy.is_reviewer}
        onChange={onChange('is_reviewer')}
      />
      <Typography style={{}}>Editor:</Typography>
      <Checkbox
        checked={workingCopy.is_editor}
        onChange={onChange('is_editor')}
      />
      <Typography style={{}}>Admin:</Typography>
      <Checkbox
        checked={workingCopy.is_admin}
        onChange={onChange('is_admin')}
      />
      <IconButton
        disabled={savingRoles || !areRolesDifferent(user, workingCopy)}
        onClick={onSave}
      >
        <SaveIcon />
      </IconButton>
    </div>
  );
};