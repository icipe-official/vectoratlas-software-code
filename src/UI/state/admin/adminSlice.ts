import { createSlice } from '@reduxjs/toolkit';
import { updateUserRoles } from '../../api/queries';
import { UsersWithRoles } from '../state.types';

export interface AdminState {
  users: UsersWithRoles[];
  loading: boolean;
  savingUser: boolean;
}

export const initialState: () => AdminState = () => ({
  users: [],
  loading: false,
  savingUser: false
});

export const adminSlice = createSlice({
  name: 'news',
  initialState: initialState(),
  reducers: {
    usersWithRoles(state, action) {
      state.users = action.payload;
    },
    adminLoading(state, action) {
      state.loading = action.payload;
    },
    savingRoles(state, action) {
      state.savingUser = action.payload;
    },
    changeUserRoles(state, action) {
      const matchingUser = state.users.find(u => u.auth0_id === action.payload.auth0_id);
      if (matchingUser) {
        matchingUser.is_admin = action.payload.is_admin;
        matchingUser.is_uploader = action.payload.is_uploader;
        matchingUser.is_reviewer = action.payload.is_reviewer;
        matchingUser.is_editor = action.payload.is_editor;
      }
    }
  },
  extraReducers: () => {},
});

export const {
  usersWithRoles,
  adminLoading,
  savingRoles,
  changeUserRoles
} = adminSlice.actions;

export default adminSlice.reducer;
