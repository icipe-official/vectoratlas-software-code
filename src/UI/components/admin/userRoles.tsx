import CircularProgress from '@mui/material/CircularProgress';
import Typography from '@mui/material/Typography';
import React, { useEffect } from 'react';
import { getUserRoles } from '../../state/admin/actions/admin.actions';
import { useAppDispatch, useAppSelector } from '../../state/hooks';
import { UserControl } from './userControl';

export const UserRolePanel = () => {
  const dispatch = useAppDispatch();
  const loadingUserRoles = useAppSelector((s) => s.admin.loading);

  const userRoles = useAppSelector((s) => s.admin.users);

  useEffect(() => {
    dispatch(getUserRoles());
  }, [dispatch]);

  return (
    <div>
      {loadingUserRoles ? (
        <div
          style={{ width: '100%', display: 'flex', justifyContent: 'center' }}
        >
          <CircularProgress />
        </div>
      ) : null}
      <Typography
        color="primary"
        variant="h5"
        sx={{ mt: 2, mb: 1 }}
        style={{ flexGrow: 1 }}
      >
        User Roles
      </Typography>
      <div>
        {userRoles.map((u) => {
          return <UserControl key={u.auth0_id} user={u} />;
        })}
      </div>
    </div>
  );
};
