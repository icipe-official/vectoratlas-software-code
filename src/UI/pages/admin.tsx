import { Container, Typography } from '@mui/material';
import { UserRolePanel } from '../components/admin/userRoles';
import AuthWrapper from '../components/shared/AuthWrapper';

export default function SourcesPage(): JSX.Element {
  return (
    <div>
      <main>
        <Container>
          <AuthWrapper role="admin">
            <div>
              <Typography
                color="primary"
                variant="h4"
                sx={{ mt: 2, mb: 1 }}
                style={{ flexGrow: 1 }}
              >
                Administration
              </Typography>
              <UserRolePanel />
            </div>
          </AuthWrapper>
        </Container>
      </main>
    </div>
  );
}
