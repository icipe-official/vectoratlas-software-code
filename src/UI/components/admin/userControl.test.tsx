import { initialState } from "../../state/admin/adminSlice";
import { AppState } from "../../state/store";
import { areRolesDifferent, UserControl } from "./userControl"
import { fireEvent, render } from '../../test_config/render';

jest.mock('../../state/admin/actions/admin.actions', () => ({
  saveUserRoles: (c) => ({type: 'saveUserRoles-mock', payload: c})
}))

const buildUser = () => ({
  email: 'test user',
  auth0_id: 'user123',
  is_admin: false,
  is_uploader: false,
  is_reviewer: false,
  is_editor: false,
});

describe('areRolesDifferent', () => {
  it('returns true if any roles are different', () => {
    const user = buildUser();
    expect(areRolesDifferent(user, user)).toBeFalsy();

    let workingCopy = buildUser();
    workingCopy.is_admin = true;
    expect(areRolesDifferent(user, workingCopy)).toBeTruthy();

    workingCopy = buildUser();
    workingCopy.is_uploader = true;
    expect(areRolesDifferent(user, workingCopy)).toBeTruthy();

    workingCopy = buildUser();
    workingCopy.is_reviewer = true;
    expect(areRolesDifferent(user, workingCopy)).toBeTruthy();

    workingCopy = buildUser();
    workingCopy.is_editor = true;
    expect(areRolesDifferent(user, workingCopy)).toBeTruthy();
  })
})

describe('UserControl', () => {
  let state: Partial<AppState>;
  const user = buildUser();

  beforeEach(() => {
    state = {
      admin: initialState()
    }
  });

  it('renders correctly', () => {
    const { wrapper } = render(<UserControl user={user}/>, state);
    expect(wrapper.container).toMatchSnapshot();
  });

  it('disables the save button if in the middle of saving', () => {
    state.admin.savingUser = true;
    const { wrapper } = render(<UserControl user={user}/>, state);

    expect(wrapper.getByRole('button')).toBeDisabled();
  })

  it('dispatches the right save action when clicked', () => {
    const { wrapper, store } = render(<UserControl user={user}/>, state);

    const checkboxes = wrapper.getAllByRole('checkbox');

    // check the editor role
    fireEvent.click(checkboxes[2]);

    // click save
    fireEvent.click(wrapper.getByRole('button'));

    const actions = store.getActions();
    const expectedUser = buildUser();
    expectedUser.is_editor = true;
    expect(actions[0]).toEqual({type: 'saveUserRoles-mock', payload: expectedUser});
  })
})