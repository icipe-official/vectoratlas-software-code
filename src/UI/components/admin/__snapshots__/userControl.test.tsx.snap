// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserControl renders correctly 1`] = `
<div>
  <div
    style="display: flex; width: 100%; align-items: center; margin-bottom: 20px;"
  >
    <div
      style="flex-grow: 1; align-items: center;"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        style="font-weight: bold; margin-right: 5px;"
      >
        test user
      </p>
      <span
        class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
      >
        (
        user123
        )
      </span>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      Uploader:
    </p>
    <span
      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
    >
      <input
        class="PrivateSwitchBase-input css-1m9pwf3"
        data-indeterminate="false"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CheckBoxOutlineBlankIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </span>
    <p
      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      Reviewer:
    </p>
    <span
      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
    >
      <input
        class="PrivateSwitchBase-input css-1m9pwf3"
        data-indeterminate="false"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CheckBoxOutlineBlankIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </span>
    <p
      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      Editor:
    </p>
    <span
      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
    >
      <input
        class="PrivateSwitchBase-input css-1m9pwf3"
        data-indeterminate="false"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CheckBoxOutlineBlankIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </span>
    <p
      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
    >
      Admin:
    </p>
    <span
      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
    >
      <input
        class="PrivateSwitchBase-input css-1m9pwf3"
        data-indeterminate="false"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CheckBoxOutlineBlankIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </span>
    <button
      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SaveIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
        />
      </svg>
    </button>
  </div>
</div>
`;
